src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1108 @@
+                )?,
+            ),
+        };
+        Ok(ResponseData::AuthenticatorClientPin(response))
+    }
+
+    pub fn check_pin_auth_token(&self, hmac_contents: &[u8], pin_auth: &[u8]) -> bool {
+        check_pin_auth(&self.pin_uv_auth_token, &hmac_contents, &pin_auth)
+    }
+
+    pub fn reset(&mut self, rng: &mut impl Rng256) {
+        self.key_agreement_key = crypto::ecdh::SecKey::gensk(rng);
+        self.pin_uv_auth_token = rng.gen_uniform_u8x32();
+        self.consecutive_pin_mismatches = 0;
+        #[cfg(feature = "with_ctap2_1")]
+        {
+            self.permissions = 0;
+            self.permissions_rp_id = None;
+        }
+    }
+
+    pub fn process_hmac_secret(
+        &self,
+        hmac_secret_input: GetAssertionHmacSecretInput,
+        cred_random: &Option<Vec<u8>>,
+    ) -> Result<Vec<u8>, Ctap2StatusCode> {
+        let GetAssertionHmacSecretInput {
+            key_agreement,
+            salt_enc,
+            salt_auth,
+        } = hmac_secret_input;
+        let pk: crypto::ecdh::PubKey = CoseKey::try_into(key_agreement)?;
+        let shared_secret = self.key_agreement_key.exchange_x_sha256(&pk);
+        // HMAC-secret does the same 16 byte truncated check.
+        if !check_pin_auth(&shared_secret, &salt_enc, &salt_auth) {
+            // Hard to tell what the correct error code here is.
+            return Err(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_EXTENSION);
+        }
+
+        match cred_random {
+            Some(cr) => encrypt_hmac_secret_output(&shared_secret, &salt_enc[..], cr),
+            // This is the case if the credential was not created with HMAC-secret.
+            None => Err(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_EXTENSION),
+        }
+    }
+
+    #[cfg(feature = "with_ctap2_1")]
+    pub fn has_make_credential_permission(&mut self, rp_id: &str) -> Result<(), Ctap2StatusCode> {
+        self.has_permission(0x01, rp_id)
+    }
+
+    #[cfg(feature = "with_ctap2_1")]
+    pub fn has_get_assertion_permission(&mut self, rp_id: &str) -> Result<(), Ctap2StatusCode> {
+        self.has_permission(0x02, rp_id)
+    }
+
+    // TODO(kaczmarczyck) use permissons for new commands
+
+    #[cfg(feature = "with_ctap2_1")]
+    fn has_permission(&mut self, bitmask: u8, rp_id: &str) -> Result<(), Ctap2StatusCode> {
+        if let Some(permissions_rp_id) = &self.permissions_rp_id {
+            if rp_id != permissions_rp_id {
+                return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);
+            }
+        } else {
+            self.permissions_rp_id = Some(String::from(rp_id));
+        }
+        if self.permissions & bitmask != 0 {
+            Ok(())
+        } else {
+            Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID)
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crypto::rng256::ThreadRng256;
+
+    fn set_standard_pin(persistent_store: &mut PersistentStore) {
+        let mut pin = [0x00; 64];
```

Comment:
In `0x00` the type is not obvious. It would be clearer to write is as `0u8` or `0u32` depending on what it actually is.